src/main/java/adris/altoclef/tasks/misc/GetToOuterEndIslandsTask.java

```diff
@@ -0,0 +1,120 @@
+import adris.altoclef.util.Dimension;
+import adris.altoclef.util.ItemTarget;
+import adris.altoclef.util.baritone.GoalAnd;
+import adris.altoclef.util.helpers.StorageHelper;
+import adris.altoclef.util.helpers.WorldHelper;
+
+import baritone.api.pathing.goals.GoalComposite;
+import baritone.api.pathing.goals.GoalGetToBlock;
+import baritone.api.pathing.goals.GoalYLevel;
+import net.minecraft.block.Blocks;
+import net.minecraft.item.Items;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.util.math.Vec3i;
+
+public class GetToOuterEndIslandsTask extends Task {
+    private Task _beatTheGame;
+    private boolean _pathingToGateway = false;
+    public final Vec3i[] OFFSETS = {
+            new Vec3i(1, -1, 1),
+            new Vec3i(1, -1, -1),
+            new Vec3i(-1, -1, 1),
+            new Vec3i(-1, -1, -1),
+            new Vec3i(2, -1, 0),
+            new Vec3i(0, -1, 2),
+            new Vec3i(-2, -1, 0),
+            new Vec3i(0, -1, -2)
+    };
+    public GetToOuterEndIslandsTask() {
+
+    }
+
+    @Override
+    protected void onStart(AltoClef mod) {
+        mod.getBehaviour().push();
+        mod.getBlockTracker().trackBlock(Blocks.END_GATEWAY);
+        _beatTheGame = new BeatMinecraft2Task();
+    }
+
+    @Override
+    protected Task onTick(AltoClef mod) {
+        if (isFinished(mod)) {
+            setDebugState("What?");
+            return null;
+        }
+        if (mod.getBlockTracker().anyFound(Blocks.END_GATEWAY)) {
+            if (!mod.getItemStorage().hasItemInventoryOnly(Items.ENDER_PEARL)) {
+                setDebugState("Getting an ender pearl");
+                return new CataloguedResourceTask(new ItemTarget(Items.ENDER_PEARL, 1));
+            }
+            BlockPos gateway = mod.getBlockTracker().getNearestTracking(Blocks.END_GATEWAY).get();
+            int blocksNeeded = Math.abs(mod.getPlayer().getBlockY() - gateway.getY()) +
+                    Math.abs(mod.getPlayer().getBlockX() - gateway.getX()) +
+                    Math.abs(mod.getPlayer().getBlockZ() - gateway.getZ()) - 1;
+            if (StorageHelper.getBuildingMaterialCount(mod) < blocksNeeded) {
+                setDebugState("Getting building materials");
+                return new GetBuildingMaterialsTask(blocksNeeded);
+            }
+            GoalAnd goal = makeGoal(gateway);
+            Debug.logMessage(mod.getPlayer().getBlockPos().toString());
+            if (!goal.isInGoal(mod.getPlayer().getBlockPos()) || !mod.getPlayer().isOnGround()) {
+                mod.getClientBaritone().getCustomGoalProcess().setGoal(goal);
+                if (!mod.getClientBaritone().getPathingBehavior().isPathing()) {
+                    mod.getClientBaritone().getCustomGoalProcess().path();
+                }
+                setDebugState("Getting close to gateway...");
+                return null;
+            }
+            setDebugState("Throwing the pearl inside");
```

Comment:
Come to think of it, why don't we throw the portal in from the ground? Why do we build to it at all? Altoclef is very good at calculating trajectories. I would suggest playing into its strengths to avoid building entirely if you can.

Reply:
I was thinking this, but I wasn't sure on how to tell if we can even hit the gateway from the position we are at

Reply:
Yeah there is no "parabola raycast" in yet (maybe should add one but oh well), so what you're doing is valid.

When the bot throws an ender pearl for the dragon fight it assumes there's nothing in the way and that it always will have a clear shot. Here it's different.